libs/ledger-live-common/src/wallet-api/logic.ts

```diff
@@ -229,3 +231,137 @@ export const bitcoinFamillyAccountGetXPubLogic = (
   tracking.bitcoinFamillyAccountXpubSuccess(manifest);
   return Promise.resolve(account.xpub);
 };
+
+export function startExchangeLogic(
+  { manifest, tracking }: WalletAPIContext,
+  exchangeType: "SWAP" | "SELL" | "FUND",
+  uiNavigation: (exchangeType: "SWAP" | "SELL" | "FUND") => Promise<string>
+): Promise<string> {
+  tracking.startExchangeRequested(manifest);
+
+  return uiNavigation(exchangeType);
+}
+
+export type CompleteExchangeRequest = {
+  provider: string;
+  fromAccountId: string;
+  toAccountId?: string;
+  transaction: WalletAPITransaction;
+  binaryPayload: string;
+  signature: string;
+  feesStrategy: string;
+  exchangeType: number;
+};
+export type CompleteExchangeUiRequest = {
+  provider: string;
+  exchange: Exchange;
+  transaction: WalletAPITransaction;
+  binaryPayload: string;
+  signature: string;
+  feesStrategy: string;
+  exchangeType: number;
+};
+
+export function completeExchangeLogic(
+  { manifest, accounts, tracking }: WalletAPIContext,
+  {
+    provider,
+    fromAccountId,
+    toAccountId,
+    transaction,
+    binaryPayload,
+    signature,
+    feesStrategy,
+    exchangeType,
+  }: CompleteExchangeRequest,
+  uiNavigation: (request: CompleteExchangeUiRequest) => Promise<string>
+): Promise<string> {
+  tracking.completeExchangeRequested(manifest);
+
+  const realFromAccountId = getAccountIdFromWalletAccountId(fromAccountId);
+  if (!realFromAccountId) {
+    return Promise.reject(new Error(`accountId ${realFromAccountId} unknown`));
```

Comment:
Do we want to leak the realFromAccountId in the error ?

Reply:
Oh wow nice catch